app/repositories/ArrivalMovementRepository.scala

```diff
@@ -145,11 +148,38 @@ class ArrivalMovementRepository @Inject()(cc: ControllerComponents, mongo: React
     )
 
     collection.flatMap {
-      _.update(false).one(selector, modifier).map {
-        y =>
-          if (y.n == 1) Some(())
-          else None
-      }
+      _.update(false)
+        .one(selector, modifier)
+        .map {
+          y =>
+            if (y.n == 1) Some(())
+            else None
+        }
+    }
+  }
+
+  def setArrivalStateAndMessageState(arrivalId: ArrivalId,
+                                     messageId: MessageId,
+                                     arrivalState: ArrivalState,
+                                     messageState: MessageState): Future[Option[Unit]] = {
+
+    val selector = Json.obj("_id" -> arrivalId)
+
+    val modifier = Json.obj(
+      "$set" -> Json.obj(
+        s"messages.${messageId.index}.state" -> messageState.toString,
+        "state"                              -> arrivalState.toString
+      )
+    )
+
+    collection.flatMap {
+      _.update(false)
+        .one(selector, modifier)
+        .map {
+          y =>
+            if (y.n == 1) Some(())
```

Comment:
This is an interesting one, in Mongo, a document can be selected for an update but not updated because the operation would set it to the current value, so in that case, `n` will be 1 and `nModified` would be 0. So we check here that there has been exactly 1 match for the update which tells us the operation has been successful